### PR TITLE
gather function only

### PR DESCRIPTION
**Pull request by zchen0211**
I add the gather functions, with unit-test passed.

The function is very similar to https://www.tensorflow.org/api_docs/python/tf/gather.

### CODE REVIEW

File: paddle/operators/gather_func.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Return a new tensor from source tensor, gathered according to index
+ * input[src]: type-T source Tensor
+ * input[Index]: type-int index Tensor (1-D)
+ * return: output tensor
+ */
+template <typename place, typename T>
+Tensor* Gather_func(Tensor* Src, Tensor* Index) {
+	// assert index is an int-type tensor?
```
Comment (wangkuiyi) [inline paddle/operators/gather_func.h]: Remove unnecessary code instead commenting them out.

File: paddle/operators/gather_func.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#include "paddle/platform/place.h"
+#include "paddle/framework/ddim.h"
+
+/**
+ * Return a new tensor from source tensor, gathered according to index
+ * input[src]: type-T source Tensor
+ * input[Index]: type-int index Tensor (1-D)
+ * return: output tensor
+ */
+template <typename place, typename T>
```
Comment (wangkuiyi) [inline paddle/operators/gather_func.h]: place ==> Place, 

We are following Google C++ style, which requires that type names in Camel case: https://google.github.io/styleguide/cppguide.html#Type_Names

File: paddle/operators/gather_func.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+Tensor* Gather_func(Tensor* Src, Tensor* Index) {
+	// assert index is an int-type tensor?
+	// assert(Index->istype(int));
+
+	// check index of shape 1-D
+	assert(Index->dims().size()==1);
```
Comment (wangkuiyi) [inline paddle/operators/gather_func.h]: Please use PADDLE_ENFORE defined at https://github.com/PaddlePaddle/Paddle/blob/6824c09d5d03545840d0278dc984d1c6bfc8c478/paddle/platform/enforce.h#L155

File: paddle/operators/gather_func.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	int index_size = Index->dims()[0];
+
+	// Source shape
+	auto src_dims = Src->dims();
```
Comment (wangkuiyi) [inline paddle/operators/gather_func.h]: The indentation doesn't match Google C++ style. This error implies that you are not using pre-commit to automatically reformat your code.

Please run `pip install pre-commit` to install pre-commit. Then Paddle repo's configuration will enable Git to call it whenever you run `git commit`.

File: paddle/operators/gather_func.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	DDim output_dims(dims_src);
+	// Create a tensor of shape [index_size, dim_src[1:]]
+	output_dims[0] = index_size;
+
+	Tensor* New_tensor;
+	float* output = nullptr;
+
+	/* slice size */
+	int slice_size = 1;
+	for(unsigned int i = 0; i < src_dims.size(); ++i)
```
Comment (wangkuiyi) [inline paddle/operators/gather_func.h]: unsigned int ==> size_t, which is the return type of `src_dims.size()`.

File: paddle/operators/gather_func.h [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+#include "paddle/framework/ddim.h"
+
+/**
+ * Return a new tensor from source tensor, gathered according to index
+ * input[src]: type-T source Tensor
+ * input[Index]: type-int index Tensor (1-D)
+ * return: output tensor
+ */
+template <typename place, typename T>
+Tensor* Gather_func(Tensor* Src, Tensor* Index) {
```
Comment (wangkuiyi) [inline paddle/operators/gather_func.h]: Src => src
Index => index

Please follow https://google.github.io/styleguide/cppguide.html#Variable_Names when naming variable.

Comment (wangkuiyi) [inline paddle/operators/gather_func.h]: Gather_func => Gather

Please follow https://google.github.io/styleguide/cppguide.html#Function_Names when naming functions.

File: paddle/operators/scatter_func.h [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+/* Implementation of CPU copy */
+template<typename T>
+void CPUScatterUpdate(const T* src, const int* Index, 
+			   const int slice_size, const int index_size,
+			   T* output) {
+  //const size_t slice_bytes = slice_size * sizeof(T);
+
+  for(int i = 0; i < index_size; ++i)
+  	int index_ = index[i];
+  	/* dst[index_] += src[index_]
```
Comment (wangkuiyi) [inline paddle/operators/scatter_func.h]: Remove unnecessary code instead of commenting them out.

File: paddle/operators/gather_func.h [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+ */
+template <typename Place, typename T>
+Tensor* Gather(Tensor* src, Tensor* index) {
+  // check index of shape 1-D
+  PADDLE_ENFORCE(index->dims().size() == 1);
+  int index_size = index->dims()[0];
+
+  // Source shape
+  auto src_dims = src->dims();
+  DDim output_dims(dims_src);
```
Comment (jacquesqiao) [inline paddle/operators/gather_func.h]: dims_src  ==> src_dims ?

File: paddle/operators/gather_func.h [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  // Create a tensor of shape [index_size, dim_src[1:]]
+  output_dims[0] = index_size;
+
+  Tensor* New_tensor;
```
Comment (jacquesqiao) [inline paddle/operators/gather_func.h]:
```cpp
Tensor* New_tensor = new Tensor();  
```
use shared_ptr<Tensor> maybe better

**Comment by zchen0211** [comment]
Function looks like this one.
gather

File: paddle/operators/gather.h [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  const size_t slice_bytes = slice_size * sizeof(T);
+
+  for (size_t i = 0; i < index_size; ++i) {
+    int index_ = indices[i];
+    // copy src[index_] to output[i]
+    memcpy(output + i * slice_size, params + index_ * slice_size, slice_bytes);
+  }
+}
+
+/* Implementation of GPU copy:
```
Comment (wangkuiyi) [inline paddle/operators/gather.h]: `/*` ==> `//`

File: paddle/operators/gather.h [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  // slice size
+  int slice_size = 1;
+  for (size_t i = 1; i < src_dims.size(); ++i) slice_size *= src_dims[i];
+
+  // Gathering
+  if (platform::is_cpu_place(place)) {
+    CPUGather<T>(src->data<T>(), index->data<int>(), slice_size, index_size,
+                 output->data<T>());
+  } else {
+    // init for GPU
```
Comment (wangkuiyi) [inline paddle/operators/gather.h]: delete unwanted  code

File: paddle/operators/gather.h [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+#pragma once
+#include <memory.h>
+#include <cstring>
+
+#include "paddle/framework/ddim.h"
+#include "paddle/framework/tensor.h"
+#include "paddle/platform/place.h"
+
+using paddle::framework::Tensor;
```
Comment (wangkuiyi) [inline paddle/operators/gather.h]: According to Google C++ style guide https://google.github.io/styleguide/cppguide.html#Namespaces, no `using` in header files.

File: paddle/operators/gather_test.cc [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+TEST(Gather, GatherData) {
+  using namespace paddle::framework;
+  using namespace paddle::platform;
+  using namespace paddle::operators;
+
+  Tensor* src = new Tensor();
+  Tensor* index = new Tensor();
+  Tensor* output = new Tensor();
+  // src.Resize(make_ddim({3, 4}));
```
Comment (wangkuiyi) [inline paddle/operators/gather_test.cc]: unwanted code

File: paddle/operators/gather_test.cc [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+  int* p_src = nullptr;
+  int* p_index = nullptr;
+  p_src = src->mutable_data<int>(make_ddim({3, 4}), CPUPlace());
+  p_index = index->mutable_data<int>(make_ddim({2}), CPUPlace());
+
+  for (size_t i = 0; i < 12; ++i) p_src[i] = i;
+  p_index[0] = 1;
+  p_index[1] = 0;
+
+  // gather
```
Comment (wangkuiyi) [inline paddle/operators/gather_test.cc]: inaccurate comment

File: paddle/operators/gather.h [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  }
+}
+
+/* Implementation of GPU copy:
+  I suppose the GPUDevice& d, contains gpu_id and thread_id
+  d = cuda_stream(gpu_id_, stream_id_);
+*/
+template <typename T>
+void GPUGather(const T* src, const int* index, const int slice_size,
+               const int index_size, T* output);
```
Comment (wangkuiyi) [inline paddle/operators/gather.h]: Do you plan to implement this function template in a future PR?

File: paddle/operators/gather.h [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
```
Comment (wangkuiyi) [inline paddle/operators/gather.h]: We can just use `//`

**Pull request merged**